File: 053 Maximum Subarray.c
/*
Maximum Subarray
https://leetcode.com/problems/maximum-subarray/

Given an integer array nums, find the contiguous subarray (containing at least one number) which
has the largest sum and return its sum.

Example:

Input: [-2,1,-3,4,-1,2,1,-5,4],
Output: 6
Explanation: [4,-1,2,1] has the largest sum = 6.
*/
/*
Kadane’s Algorithm
Assume $size >= 1
And notice $global CANNOT start from 0
*/
int kadaneMax(int* arr, int size)
{
    int local, global, i;
    if(size == 0) return -1;

    local = arr[0];
    global = local;

    for(i = 1;i<size;i++)
    {
        if(local < 0) local = 0;
        
        local += arr[i];
        if(global < local) global = local;
    }
    return global;
}  
int kadaneMin(int* arr, int size)
{
    int local, global, i;
    if(size == 0) return -1;

    local = arr[0];
    global = local;

    for(i = 1;i<size;i++)
    {
        if(local > 0) local = 0;
        
        local += arr[i];
        if(global > local) global = local;
    }
    return global;
}

File: 053 Maximum Subarray.ts
/*
53. Maximum Subarray
https://leetcode.com/problems/maximum-subarray/

Example 1:

Input: nums = [-2,1,-3,4,-1,2,1,-5,4]
Output: 6
Explanation: The subarray [4,-1,2,1] has the largest sum 6.

Example 2:

Input: nums = [1]
Output: 1
Explanation: The subarray [1] has the largest sum 1.

Example 3:

Input: nums = [5,4,-1,7,8]
Output: 23
Explanation: The subarray [5,4,-1,7,8] has the largest sum 23.

Constraints:

    1 <= nums.length <= 105
    -104 <= nums[i] <= 104
    
*/
function maxSubArray(nums: number[]): number {
    let result = nums[0];
    let total = nums[0];
    for (let i = 1; i < nums.length; i++) {
        total = Math.max(total + nums[i], nums[i]);
        result = Math.max(result, total);
    }
    return result;
}

